src/charm.py

```diff
@@ -189,22 +200,76 @@ def _on_leader_elected(self, _) -> None:
                 )
         self.unit_peer_data.update({"leader": "true"})
 
+        # Create and set cluster and cluster-set names in the peer relation databag
+        common_hash = generate_random_hash()
+        self.app_peer_data.setdefault(
+            "cluster-name", self.config.cluster_name or f"cluster-{common_hash}"
+        )
+        self.app_peer_data.setdefault("cluster-set-domain-name", f"cluster-set-{common_hash}")
+
     def _on_leader_settings_changed(self, _) -> None:
         """Handle the leader settings changed event."""
         self.unit_peer_data.update({"leader": "false"})
 
-    def _on_config_changed(self, _) -> None:
+    def _on_config_changed(self, event: EventBase) -> None:
         """Handle the config changed event."""
-        # Only execute on leader unit
-        if not self.unit.is_leader():
+        if not self._is_peer_data_set:
+            # skip when not initialized
             return
 
-        # Create and set cluster and cluster-set names in the peer relation databag
-        common_hash = generate_random_hash()
-        self.app_peer_data.setdefault(
-            "cluster-name", self.config.get("cluster-name", f"cluster-{common_hash}")
+        if not self.upgrade.idle:
+            # skip when upgrade is in progress
+            # the upgrade already restart the daemon
+            return
+
+        if not self._mysql.is_mysqld_running:
```

Comment:
this is not a property. should we add `()` to invoke the method?

Reply:
740b58e